### DISCUSSION TITLE
Challenges with setting up Infracost

### DISCUSSION THREAD

**Discussion by alikhajeh1** [infracost/infracost#1339] [Q&A]
# Latest update

👉👉👉 The top 3 problems mentioned in this discussion have been solved with the **Infracost v0.10 release**.
👉👉👉 Checkout [the migration guide](https://www.infracost.io/docs/guides/v0.10_migration/) for details.

--- 

**Challenges with setting up Infracost**

We're investigating challenges that users have setting up Infracost. We would love for you to comment on (or vote) on the following questions. You can also add other problems you ran into so we can discuss it with the rest of the community.

**Comment by alikhajeh1** [comment]
### 1. Is having access to cloud credentials an issue for you?

Terraform needs access to cloud credentials to generate a plan JSON file (which Infracost uses to generate a cost estimate).

Has this been an issue for you? If yes, are you ok with a potentially less accurate cost estimate? or would you rather go through the pain of setting up cloud creds to get a more accurate estimate?

**Reply by AErmie** [reply]
For some customers it could be an additional challenge to set up a pipeline with cloud credentials, etc. I'd like to see a default option as the less-accurate estimate, when a mention/notice of "for a more accurate estimate, please use cloud credentials to create a JSON terraform plan file."

**Reply by brettcurtis** [reply]
We are running Infracost in the same Action workflows that build the infra so credentials are not a problem.

**Reply by andylw-cap** [reply]
I would suggest having the options for both with and without credentials for those who may require it - be it in a demo or otherwise.

**Comment by alikhajeh1** [comment]
### 2. Is having to install the terraform/terragrunt binaries causing you pain?

If yes, is it a pain in on your laptop or in CI/CD?

If it's a pain in CI/CD, I assume you did not already run terraform there?

**Reply by AErmie** [reply]
It's more of an issue (in my experience) for CI/CD, especially if it's a shared build agent/server environment. This is why I always opt to use a Docker container approach, so it's encapsulated.

For tools (like InfraCost) that require a terraform plan file, I end up running terraform init/plan in a Docker container, and capture the output to be able to pass it into another Docker container (with the target tool, like InfraCost).

**Reply by brettcurtis** [reply]
Same here, our credentials are available to the workflows already so not a problem.

**Comment by alikhajeh1** [comment]
### 3. Is running Infracost (which can trigger terraform init/plan) too slow for you?

If yes, is it it too slow in the CLI, CI, or both? What does "too slow" mean for you?

**Reply by AErmie** [reply]
"Slow" is relative, of course. I'm curious about the internal workings of InfraCost. There's another tool that I've used (called [MegaLinter](https://megalinter.github.io/)), that built on the idea of the [GitHub Super-Linter](https://github.com/github/super-linter), but improved it by running scans in parallel instead of sequentially.

I'm curious if InfraCost does any parallelization of it's execution (if applicable).

**Reply by alikhajeh1** [reply]
Thanks for the awesome replies @AErmie! The CLI retrieves the cloud prices in parallel; also when a config-file is used then the projects in it are processed in parallel. That has provided significant speed-ups so far. When using the Infracost CLI outside of CI/CD, it caches the Terraform plan for ~30mins (invalidated automatically if something changes) but it's still not as fast as linting tools due to terraform init/plan taking a while...

**Reply by brettcurtis** [reply]
We've only poked at Infracost for module development and we keep our modules as small as possible so it's been fine there. I'm going to be testing Infracost on a more complex infrastructure build-out involving several different child modules as well as multi-region pretty soon and will report back with any concerns.

**Comment by alikhajeh1** [comment]
### Add your problems!

so we can discuss it in the community

**Reply by AErmie** [reply]
I know that we've already discussed this previously, but I thought I'd add it here to start the community conversation...

Not everyone uses Terraform, so it would be great if InfraCost supported platform-native languages (ie. ARM and Bicep for Azure, CloudFormation for AWS, the equivalent for GCP).

**Reply by AErmie** [reply]
Further, for enterprise customers, it would be more appealing if InfraCost could utilize their specific cloud pricing (ie. enterprise discounts, CSP, etc.). That way InfraCost would provide a more realistic estimate, versus pricing based off of the "list" price for the public cloud provider.

**Reply by dfrkp** [reply]
I'd add the Integration with Terraform Cloud. Specifically, when one uses the VCS integration of Terraform Cloud. In this case I have to run (trigger) an additional run from the CI only to then parse for the plan id so it can be fetched again. This burns CI minutes and slows down the whole process.

**Reply by alikhajeh1** [reply]
@dfrkp are you referring to [TFC Run Tasks](https://www.infracost.io/docs/integrations/terraform_cloud_enterprise/#terraform-cloud-run-tasks) or running the CLI manually (which does something [like this](https://www.infracost.io/docs/troubleshooting/#terraform-cloud) internally)?

**Reply by dfrkp** [reply]
@alikhajeh1 thabks for pointing out the cloud run tasks! That looks promising! I was referring to the second link of yours which would start an additional plan in parallel of you use the external pipeline feature with terraform cloud.

**Comment by alikhajeh1** [comment]
Hey everyone, we're experimenting with parsing HCL to see if we can make it simpler to run Infracost on the CLI/laptops and open-up new use-cases. We'd appreciate it if you [try this](https://www.infracost.io/docs/guides/advanced_usage/#parsing-hcl-experimental) and create GitHub issues if you find any discrepancies.

**Comment by alikhajeh1** [comment]
Hi everyone, the top 3 problems mentioned in this discussion have been solved with the Infracost v0.10 release, checkout [the migration guide](https://www.infracost.io/docs/guides/v0.10_migration/) for details.

Thank you for helping!
